enhancements/installer/custom-dns.md

```diff
@@ -0,0 +1,574 @@
+(ELB). This part of the solution can be worked on in parallel and combined later.
+
+
+### Risks and Mitigations
+
+OpenShift Installer for the AWS platform (maybe other platforms too) currently
+configures the Load Balancer first. After the LB configuration completes
+successfully, we are returned a aws_lb_api_external_dns_name which is used
+for the DNS configuration. The e2e solution for this custom DNS solution should
+account for the fact that the customer would be asked to configure their own DNS
+entries when the Load Balancers that these entries point to may not be easily
+known to them.
+
+The e2e solution should do its best to help the customer figure out the DNS
+configuration that they are responsible for without making the workflow too
+dependent on user intervention.
+
+### Drawbacks
+
+The idea is to find the best form of an argument why this enhancement should
+_not_ be implemented.  
+
+What trade-offs (technical/efficiency cost, user experience, flexibility, 
+supportability, etc) must be made in order to implement this? What are the reasons
+we might not want to undertake this proposal, and how do we overcome them?  
+
+[Note: No reason has emerged so far to indicate that this should not implemented.]
+
+Does this proposal implement a behavior that's new/unique/novel? Is it poorly
+aligned with existing user expectations?  Will it be a significant maintenance
+burden?  Is it likely to be superceded by something else in the near future?
+
+[Note: Decision made in the Design Details section can affect our answer to this
+question.]
+
+## Design Details
+
+### Open Questions [optional]
+
+One of the decisions that needs to be made early in the design process is to
+determine the best way to order OpenShift's configuration of the Load Balancer
+and the customer's configuration on their custom DNS solution.
+
+Today, the Installer configures the LB first and the generated DNS name is used
+to configure DNS CNAME records. The LB DNS name cannot be predicted in advance
+because it is generated using an unique UUID while configuring LB entries for
+API and Ingress services.
+
+Here are a few options being considered as potential solutions:
+
+#### Option 1: Add a pre-install step to configure LB and provide DNS config
+
+This solution considers adding a pre-install step that configures Load Balancers
+for API and Ingress services. The DNS names generated per subnet/Availability Zone
+for API and Ingress services are output to the user in the form of DNS CNAME
+config. Customers can use this generated output to configure their DNS servers.
+
+This step could be run as an installer command like
+"openshift-install generate-custom-dns-config" while providing it with a complete
+install-config. After thsi step, when "create cluster" is called with the
+customDNS feature enabled, the validation step will have to ensure both the LB and
+DNS have been configured.
```

Comment:
I'm not sure I understand the suggestion here. What are the inputs and outputs? Is this step creating the load balancer infrastructure?

Reply:
I have Option 1 which talks about actually configuring the LB in this pre-install step. It will take install-config as input. Output would be DNS config that the user could use to configure custom DNS.
Option 1A would be just generate the DNS config without actually configuring the LB. The success of this option heavily depends on figuring out if there is a way to predictably generate LB DNS names. The infra-id being the only randomly generated string by the installer and it can be saved in the install-config to be re-used during "create cluster" .

Reply:
I am not attached to any of the names in this proposal. Let us name this thing based on what we finally decide its tasks are going to be. We 1st need to figure out which of the 2 Options proposed here is going to be our final solution.